---
title: " "
output: html_document
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
![](image/Image_2.jpg)

# Evolutionary Genomics & Bioinformatics
Author: "Billy (Peng) Zhou"


[Lab1:Introduction to R and Reproducible ResearchFile](Lab_1.html)

[Lab2:git, GitHub and Rstudio projectsFile](Lab2_Rmd.html)

[Lab3A:Data Wrangling:Starting with data](Lab3A.html)

[Lab3B Data Wrangling:Manipulating, analyzing and exporting data with tidyverse](Lab3B.html)

[Lab4 Data wrangling and graphing COVID-19 reporting data (Part 1)](Lab4.html) 

[Lab4A Data visualization with ggplot2 (Part 1)](Lab4a.html) 

[Lab5 Data wrangling and graphing COVID-19 reporting data (Part 2)](Lab5.html) 

[Lab5A Data visualization with ggplot2 (Part 2)](Lab5a.html) 

[Lab5 extra: Interactive Graphs and Animations from the COVID-19 reporting data](Lab5extra.html)

[Lab6 Data Maps and Interactive Graphs from the COVID-19 reporting data](Lab6.html) 

[Lab6a Plot Display](Lab6a.html)

[Lab6b R Shiny apps](Lab6b.html)

[Lab7 RNA-Seq workflow: gene-level exploratory analysis and differential expression](Lab7.html)

[Lab8: Sequence Alignment, Phylogenetic Analysis and Tree Visualization](Lab8.html)

[Lab9A : Programming in R - Regular Expressions](Lab9A.html)

[Lab9B : Programming in R - Control Structures and Functions](Lab9B.html)

[Lab10: Population genomics based on high throughput sequencing (HTS)](Lab10.html)

[Lab 11: Microbiome Analysis using dada2 and phyloseq](Lab11.html)

Application written in R [@RCoreTeam] using the Shiny framework [@Chang2015].

# REFERENCE
